src/lib/mgba/websocket-client.ts

```diff
@@ -0,0 +1,554 @@
+              this.connected = false
+              reject(error)
+            })
+
+            ;(this.ws as unknown as {
+              on: (event: string, handler: (...args: unknown[]) => void) => void
+            }).on('close', () => {
+              console.log('WebSocket connection closed')
+              this.connected = false
+            })
+          }
+        }
+
+        setupEventHandlers()
+      } catch (error) {
+        reject(error)
+      }
+    })
+  }
+
+  /**
+   * Disconnect from the WebSocket server
+   */
+  disconnect (): void {
+    if (this.ws) {
+      this.ws.close()
+      this.ws = null
+    }
+    this.connected = false
+  }
+
+  /**
+   * Check if currently connected
+   */
+  isConnected (): boolean {
+    return this.connected && this.ws?.readyState === WebSocket.OPEN
+  }
+
+  /**
+   * Execute Lua code on the mGBA emulator
+   */
+  async eval (code: string): Promise<MgbaEvalResponse> {
+    if (!this.isConnected()) {
+      throw new Error('Not connected to mGBA WebSocket server')
+    }
+
+    return new Promise((resolve, reject) => {
+      if (!this.ws) {
+        reject(new Error('WebSocket is null'))
+        return
+      }
+
+      // Set up one-time message handler for this eval
+      const messageHandler = (event: { data: unknown }) => {
+        const message = typeof event.data === 'string' ? event.data : String(event.data)
+
+        // Skip welcome messages and other non-JSON responses
+        if (message.includes('Welcome to WebSocket Eval') ||
+            message.includes('mGBA WebSocket server ready') ||
+            !message.startsWith('{')) {
+          return // Don't resolve/reject, wait for actual response
+        }
+
+        try {
+          const response = JSON.parse(message) as MgbaEvalResponse
+          if (this.ws) {
+            if ('removeEventListener' in this.ws) {
+              this.ws.removeEventListener('message', messageHandler)
+            } else if ('off' in this.ws) {
+              ;(this.ws as unknown as { off: (event: string, handler: unknown) => void }).off('message', messageHandler)
+            }
+          }
+          resolve(response)
+        } catch (error) {
+          if (this.ws) {
+            if ('removeEventListener' in this.ws) {
+              this.ws.removeEventListener('message', messageHandler)
+            } else if ('off' in this.ws) {
+              ;(this.ws as unknown as { off: (event: string, handler: unknown) => void }).off('message', messageHandler)
+            }
+          }
+          reject(new Error(`Failed to parse eval response: ${String(error)}`))
+        }
+      }
+
+      // Add event listener using appropriate method
+      if ('addEventListener' in this.ws) {
+        this.ws.addEventListener('message', messageHandler)
+      } else if ('on' in this.ws) {
+        ;(this.ws as unknown as { on: (event: string, handler: unknown) => void }).on('message', messageHandler)
```

Comment:
this line is ugly

Reply:
please make use of proper isomorphic-ws throughout this file. check the type definitions and apply how to use it correctly in this file. review this whole file for better isomorphic-ws use